### DISCUSSION TITLE
Dynamically adding forms to collection

### DISCUSSION THREAD

**Discussion by boosh** [jrief/django-formset#156] [Ideas]
I have a list of objects I want to display settings forms for. I want to avoid manually creating a form per object and instead retrieve the objects, then iterate over them instantiating a form for each one, e.g.:

```
class CampaignCollectionForm(SetProjectMixin, FormCollection):
    campaign = CampaignForm()

    def __init__(self, *args, **kwargs):
        project = kwargs.pop("project")
        
        channels = Channels.objects.filter(blah)
        for channel in channels:
            setattr(self, f"{channel.name}_form") = ChannelForm(channel=channel)
            
        super().__init__(*args, **kwargs)
```

This doesn't seem to work due to the work of `FormCollectionMeta.__new__`. Is it possible to add forms dynaically like this?

**Comment by jrief** [comment]
`__init__`probably is the wrong method for this, because then the collection class is already created. You presumably have to override the `__new__`-method in your own meta-class, or use the `type()`-function.

From looking at your code, it however seems that creating some customized initial data is the way to go.

**Reply by boosh** [reply]
How do you mean about creating customised initial data? I want to instantiate the same form multiple times for different object instances, and do all that dynamically.

**Reply by jrief** [reply]
Since you just want to initiate the same sub-form (`ChannelForm`) multiple times, why don't you use a Collection with siblings?

**Reply by boosh** [reply]
Would that allow me to create the subforms dynamically (i.e. as a result of a DB query with a varying number of results per user)?

**Reply by jrief** [reply]
Yes. If you initialize the collection (with siblings) you can pass a list of dicts as `initial`-data. The length of that list determines the number of sub-collections. The data in each dict determines the initial data of the sub-collection. Please check the docs on how to create collections with siblings.

**Reply by boosh** [reply]
OK thanks I'll try that.

**Comment by boosh** [comment]
Thanks for your help but I couldn't work it out so just stuck with standard django forms
